Support output formats "plain" (with and without colors) and "markdown"

Enselic opened the pull request:
This PR makes diffing use colors by default:
<img width="812" alt="Skärmavbild 2022-04-01 kl  17 09 04" src="https://user-images.githubusercontent.com/115040/161291288-bc7e8aed-e6f6-4c5e-8d8b-d92a627fdf22.png">

But if you pipe to a file (or pass `--color=never`) then the tool will use + and - instead of red and green:
<img width="844" alt="Skärmavbild 2022-04-01 kl  17 09 22" src="https://user-images.githubusercontent.com/115040/161291389-c3d7222d-377f-4395-a430-f9c2792eb41f.png">

douweschulte commented on src/arg_types.rs:
```diff
@@ -0,0 +1,67 @@
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "auto" {
```
Couldn't this be refactored to a match construct?
```rust
match s {
    "auto" => Ok(Color::Auto),
    "never" => Ok(Color::Never),
    "always" => Ok(Color::Always),
    _ => Err(anyhow!("See --help")),
}
```

Enselic replied:
Thanks, good catch, fix pushed

douweschulte commented on src/arg_types.rs:
```diff
@@ -0,0 +1,67 @@
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "plain" {
```
Couldn't this be refactored to a match construct?
```rust
match s {
    "plain" => Ok(OutputFormat::Plain),
    "markdown" => Ok(OutputFormat::Markdown),
    _ => Err(anyhow!("See --help")),
}
```

Enselic replied:
Thanks, good catch, fix pushed

douweschulte commented on src/markdown.rs:
```diff
@@ -1,53 +1,48 @@
+
+impl OutputFormatter for Markdown {
+    fn print_items(&self, _w: &mut dyn Write, _args: &Args, _items: Vec<PublicItem>) -> Result<()> {
+        todo!("Not yet implemented because I'm not sure what the output should look like. Feel free to open a PR with a proposal.")
```
Would a plain list just work? At least for now. Grouping related methods/structs (all struct fields as a bullet list?) would be really cool, but that should just work the same as the diff variant.

I would totally agree that this is not a very useful feature, I would value `docs.rs` above a plain list. But creating a plain list is maybe better than a panic?
```
print_items_with_header(
            w,
            "## Public API",
            &items,
            |w, item| writeln!(w, "* `{}`", item),
        )?;
```

Enselic replied:
Yep I had some kind of plain list in mind, but wasn't sure which variant to use. Now that you have cast your vote, I agree it makes good sense, so I took your code and put it in a commit with you as the author (it felt wrong to put me as author).

Grouping should indeed be handled universally by `public_items` rather than by `cargo public-items`. The issue for it is https://github.com/Enselic/public_items/issues/34. I've been trying to figure out for a while how to solve that in a good way, but it's tricky. I'm pretty sure your tokenization of `PublicItem` will make it easier to solve that.

I actually find plain lists of public items quite useful, but then I would use "plain" output format and not "markdown". I've found myself doing things like `cargo public-items | grep public_items::diff` for example.

douweschulte commented on src/output_formatter.rs:
```diff
@@ -0,0 +1,27 @@
+    fn print_diff(&self, w: &mut dyn Write, args: &Args, diff: &PublicItemsDiff) -> Result<()>;
+}
+
+pub fn print_items_with_header<T>(
```
Would this fit better as a part of the above trait? Just as a [default implemented method](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations). If you provide the implementation in the trait itself the implementers need not be changed.

Enselic replied:
Interesting idea, I tried it out, but got:

```
7  | pub trait OutputFormatter {
   |           --------------- this trait cannot be made into an object...
...
11 |     fn print_items_with_header<T>(
   |        ^^^^^^^^^^^^^^^^^^^^^^^ ...because method `print_items_with_header` has generic type parameters
```

and I think removing the generic type parameter is too big of an operation at the moment. So if you don't mind I think we can keep it as it is.

douweschulte replied:
I overlooked that then, I think it is fine where it is.

douweschulte reviewed (approved):
Looks like a nice reorganisation! I left some comments in the files.

Enselic commented:
Big thanks for code review! I addressed all comments now. As a next step I will add tests for this change before I merge it, which will require me to land some test-refactorings first, mainly turning `test-invocation-variants.sh` into proper `cargo test` tests.

douweschulte commented:
Nice! Looking forward to it!

Enselic commented:
Ok, tests added. I feel ready to merge, but I wanted you to give a chance to review my tests before merge. Doing another review round is completely optional for you of course.
